### DISCUSSION TITLE
Prisma 2 Documentation on where clause with multiple conditions in update and delete statements

### DISCUSSION THREAD

**Discussion by Dreamystify** [prisma/prisma#4185] [Q&A] [answered]
I cant seem to find any information on running the following code:
```ts
    return await prisma.users.update({
        where: {
            user_id: id,
            active: true,
        },
        ...data,
    });
```
It wont allow the 2 where clauses, and the examples are simple where most projects aren't. Is it possible to put some more complex examples in the examples repo for myself and others to shift from sql to prisma seamlessly

**Comment by ryands17** [comment, marked as answer]
@Dreamystify 👋 

We have docs on the supported operations [here](https://www.prisma.io/docs/concepts/components/prisma-client/crud). Currently you're trying to update on multiple conditions where one of them is not unique i.e. the `active` part. `update` just taken in unique values so in this case, you would need to replace `update` with `updateMany`.

**Reply by Dreamystify** [reply]
Oh ok, it works with the updateMany, but the user_id is always going to be 1 record, and the active flag is there to indicate the record is deleted or not. I was going for the record not found when it doesn't meet the condition but return the user when it does. Instead it just returns a count for the records updated. I thought when you updated a record, it returned the record.

**Reply by ryands17** [reply]
The returning of the record only works for single operations. Bulk operations like `updateMany` and `deleteMany` always return the count of records udpated similar to how SQL does.

**Reply by OzzieOrca** [reply]
Isn't specifying one unique value enough? Then we could continue scoping the query by another field? I'm interested in this feature request https://github.com/prisma/prisma/issues/1924 but for updates and deletes also.

**Reply by ryands17** [reply]
> Isn't specifying one unique value enough? Then we could continue scoping the query by another field?

Could you elaborate a bit more on this? As for the feature request you have mentioned, it already has been implemented and you can check that out in the docs [here](https://www.prisma.io/docs/reference/api-reference/prisma-client-reference#findfirst).

**Reply by OzzieOrca** [reply]
I guess there are 3 parts to this:
1. I don't understand the restriction requiring only a single condition for `findUnique`, `update`, and `delete`. Requiring at least one unique field totally makes sense. But once that requirement is met, why can't we provide more conditions that would only serve to make it more unique? Trying to do something like this
    ```
    return db.model.findUnique({
      where: { id, user: { id: context.currentUser.id } },
    });
    ```
    throws this error
    ```
    Argument where of type ModelWhereUniqueInput needs exactly one argument, but you provided id and user. Please choose one.
    ```
    To quote from the [linked feature request](https://github.com/prisma/prisma/issues/1924#issue-587238441):
    > Maybe require at least one unique property, but also allow other properties. That doesn't change the fact the findOne will always return at most a single record.
    
    (yes `findOne` is now named `findUnique`)
2. I'm totally aware the [linked feature request](https://github.com/prisma/prisma/issues/1924#issue-587238441) was implemented with `findFirst` and is in production. I'm a little confused why this support couldn't be added to `findUnique` but swapping `findUnique` for `findFirst` totally works. `findFirst` does have other use cases where none of the conditions you are filtering by are unique and you just want the first of many possible matches.
3. Using `findFirst` fixes this issue for reads but there is no similar solution for updates or deletes. Right now if you want to update or delete by multiple conditions **AND** have fields returned, you have run multiple queries. For update you have to run `updateMany` and then run `findFirst` and for delete you have to run `findFirst` and then run `deleteMany`.

**Reply by ryands17** [reply]
I think the 1st point can be achieved by `findFirst` where multiple conditions can be added and as for the 3rd point, I would suggest creating a feature request [here](https://github.com/prisma/prisma/issues/new?assignees=&labels=&template=feature_request.md&title=) so that we can look into this. Currently in case of updates and deletes for multiple items, Prisma follows the database convention of returning the count of items changed.

**Reply by yazer79** [reply]
updateMany({ where: { id, user_id}} will update all of the results having either id or user_id? Or as @OzzieOrca suggested to be associated with context as an extra security to update your own stuff only? So that would be occurrence of both id + user_id at same row

**Reply by ryands17** [reply]
This will result in an `and` so it will check if both `id` and `user_id` match.

**Reply by yazer79** [reply]
Then I think the `FindMany` name is misleading in this case

**Reply by ryands17** [reply]
It's not as multiple conditions can be used. Prisma currently doesn't know if you're combining a unique condition with a non-unique one so `findMany` could return multiple records. `findUnique` is what you're looking for if you just want to query on a unique condition.

**Reply by phoenixeliot** [reply]
I would really like if the documentation on "AND" and "OR" were easier to find. I still can't find them, and I already found them yesterday. The words aren't very searchable.

**Reply by ryands17** [reply]
@phoenixeliot 👋 
It's all documented here: https://www.prisma.io/docs/reference/api-reference/prisma-client-reference

You can go to this page from Reference section and selecting **Prisma Client API Reference**.

![image](https://user-images.githubusercontent.com/19697099/123738031-ad236d80-d8c1-11eb-82c6-668a4c3aa763.png)

**Reply by sehmaschine** [reply]
I don't fully get this restriction with `update` as well. Let's say I want to update a record by ID, but at the same time I need to check that this object belongs to the currently logged-in user. Therefore, I need two attributes `user_id` and `id` within my where clause – but that's not possible. Do I really need to wrap all update functions with a find(first)? That seems to be overly complicated IMO, but maybe I'm missing something?

**Reply by ryands17** [reply]
@sehmaschine You can use `updateMany` in that case for multiple conditions.

**Reply by sehmaschine** [reply]
Not really – because then I just get the number of results and I need an additional query to get the object (again). So it's always (at least) 2 queries for one update. Right?

**Reply by sehmaschine** [reply]
Besides, `updateMany` would silently fail if there's not matching, right?. In my opinion that`s not the behaviour one would expect with updating a single record.

**Reply by ryands17** [reply]
Yes that would be correct. In this case, you could choose whichever route you would like to take in this as currently there's no other way to query for multiple conditions in `update`.

**Reply by sehmaschine** [reply]
@ryands17 Ok, thanks for the clarification. Pretty weird decision in my opinion (but that's just my 2 cents and you probably know much better what you're doing).

**Reply by ryands17** [reply]
We do have a feature request so it would be great if you could add a 👍 and your comments so that we can look at the priority 🙂

**Reply by ridhwaans** [reply]
would `@@unique` do the job for multiple where conditions?  
example from docs 
```prisma
model TimePeriod {
  year          Int
  quarter       Int
  total         Decimal

  @@unique(fields: [year, quarter], name: "timePeriodId")
}
```
```ts
const timePeriod = await prisma.timePeriod.findUnique({
  where: {
    timePeriodId: {
      quarter: 4,
      year: 2020,
    },
  },
})
```

**Reply by ryands17** [reply]
Yes this will work.

**Reply by jaydip-github** [reply]
> Yes this will work.

sir, can we update multiple records with multiple conditions (each record has its own condition parameter), is it possible?

**Comment by brandonmpetty** [comment]
I would like to second this "issue".
For example, I am creating a REST service that uses an 'If-Unmodified-Since' header to avoid mid-air-collisions.
A great way of solving this for Delete, Patch, and Put operations would be to have a WHERE clause with my target 'id' AND then specify where 'updatedAt' is 'lte' a specified date-time.  That way it would only take the action if it were updated out from under me.

Technically, if you have a required unique field, like the primary key in my case... adding as optional any other field to filter on should be possible and you will be guaranteed that the result is unique.

Without this feature, how else am I suppose to accomplish this mid-air-collision avoidance?  By creating a transaction with an otherwise extraneous select in it?

**Comment by bdelamotte** [comment]
It would be a tremendous benefit to put multiple conditions inside the where clause of an update and delete statement. 
Use case:
I want to update a listing, but I want to ensure it is owned by the user passed in. 
I'd like to do something like this:
```
             where: {
                 id: id,
                 userId: user.id
             },
```

I get the user.id from the session so if someone is trying to update something they don't own, then the update would fail.

**Reply by atawfique21** [reply]
Does anyone have a solution to this? I'm trying to do this exactly.

**Reply by atawfique21** [reply]
> Does anyone have a solution to this? I'm trying to do this exactly.

Reading back through this thread, I saw the mention to 
> you would need to replace `update` with `updateMany`.

And tested this with `deleteMany` and it works. 

```JavaScript
await prisma.posts.deleteMany({
      where: {
        id: id,
        userId: session.user.id
}
```

This seems really hacky in a way, and `delete` itself should be able to take multiple conditions.

**Reply by nenadfilipovic** [reply]
If you dont need return value of update/delete then it is fine. But if you need it for updating cache or something else it wont work.

**Reply by CallumHYoung** [reply]
I second this! updateMany is working for me, however it is significantly more clean to only have update with multiple conditions.

**Reply by nenadfilipovic** [reply]
```ts
const [ count, returnObject ] = await prisma.$transaction([
    prisma.device.updateMany({
      where: {
        AND: composeQuery,
      },
      data,
    }),
    prisma.device.findUnique({ where }),
  ]);
```

You can do something like this if you need to return data.

**Reply by vaughngit** [reply]
Any way of returning data with deleteMany? Cant select whats been deleted.

**Reply by nenadfilipovic** [reply]
Maybe something like this

```ts
const [deletedDevice] = await prisma.$transaction([
    prisma.device.findFirst({ where }),
    prisma.device.deleteMany({ where: { AND: composeQuery } }),
  ]);
  return deletedDevice;
```

**Reply by brandonmpetty** [reply]
@nenadfilipovic I have seen those patterns proposed before.  Are you claiming that the findFirst/findUnique will put a lock on the row so that other threads trying to perform an Update, for example, will be blocked until the full transaction is done?  If true, am I correct to say that this is not working for SQLite, since PrismaJs does not claim to have transaction support for it?

**Reply by nenadfilipovic** [reply]
I am really not sure.

**Reply by 2chanhaeng** [reply]
I solve this with [`@@unique`](https://www.prisma.io/docs/reference/api-reference/prisma-schema-reference#unique-1)!
1. Add `@@unique` in the defintion of the model to update.
  ```prisma
  // prisma/schema.prisma
  model <The model to update>{
    <id field>  <id type> @id
    <not nuique field> <not nuique type>
    ...
    @@unique([<id field>, <not nuique field>]) // The most important part!
  }
  ```
2. Migrate DB.
3. Then you can update where: `<id field>_<not nuique field>`!
  ```ts
  import { PrismaClient } from "@prisma/client";
  const db = new PrismaClient();
  db.todo.update({
    where: {
      "<id field>_<not nuique field>": {
        "<id field>": "<id value>",
        "<not nuique field>": "<not nuique value>"
      },
    data: { "<update field>": "<update value>" },
    });
```
Also, you can name it!
```prisma
@@unique(..., name:"<unique name>")
```
Then you can update using it like this!
```ts
db.todo.update({
  where: {
    "<unique name>": {
        "<id field>": "<id value>",
        "<not nuique field>": "<not nuique value>"
      },
...
```
The default name is very complicate, so I recomend name the unique field!

**Comment by blevine** [comment]
It seems to me that having to do an "extra" SELECT to retrieve the data you may have just updated is not that big of a deal.  If you were not using an ORM like Prisma, this is what you'd need to do. In the updateUnique case, is Prisma just doing that SELECT for you? It looks like the confusion here boils down to the fact that one type of UPDATE returns the new data and one just returns counts. Although I agree that the fact that you can't do an updateUnique with a predicate containing multiple conditions that are guaranteed to refer to a unique row is confusing.

**Reply by vaughngit** [reply]
really?

**Reply by blevine** [reply]
yes, really

**Reply by vaughngit** [reply]
Its that kind of mindset that kills innovation. If this is the mindset of the developers, it would be good to know.

**Reply by panzi** [reply]
>  If you were not using an ORM like Prisma, this is what you'd need to do.

Not true. If I'd use the DB directly I'd do:
```SQL
UPDATE things SET foo = 'bar' WHERE id = ? AND user_id = ? RETURNING *;
```
Other ORMs support this. The nice thing about this is that this is actually atomic! `UPDATE` - `SELECT` isn't.

**Comment by blevine** [comment]
@panzi I was not aware of `RETURNING *.`  So I learned something new today.  Thank you :)

**Comment by bpetty-interlace** [comment]
Any idea of when something like this could be supported?

**Comment by bpetty-interlace** [comment]
A related issue: https://github.com/prisma/prisma/issues/7290

The reason this probably wont happen any time soon is because the prisma framework generates classes specifically for when we are doing a unique selection, to only expose those properties.  This would require deleting all of that work and basically using the same class used by the "Many" calls to expose all of the properties, with the caveat that when doing a select, update or delete... it would iterate through to make sure you chose to filter by at least one property that can determine uniqueness.  While significantly more usable, you turn what was once a compile time catch into something that will throw at run time.  Honestly, I don't know how Prisma is handling composite indexes with this... it could be a complete mess.  All that to say, you're probably better off to write a plugin that exposes this functionality than to expect anyone on the Prisma project to actually redesign everything.  This discussion is almost 2 years old already.

**Reply by janpio** [reply]
It is definitely not super high on our priority list, compared to all the other things we are working on. But generally we of course see the use case here, and also hear the need - there are a bunch of issues like that one. And we will take a look at all these sooner or later and then potentially also try to implement something. But it is hard to say when that will be unfortunately.

**Comment by doiya46** [comment]
Use `extendedWhereUnique` feature in `schema.prisma` file

```
generator client {
  provider        = "prisma-client-js"
  previewFeatures = ["extendedWhereUnique"]
}
```

**Reply by Abdel-Monaam-Aouini** [reply]
Nice :)

**Reply by anastaciom** [reply]
thanks, this was helpful and worked for me

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
